python/Dockerfile

```diff
@@ -32,43 +34,89 @@ RUN apt-get update \
+# around that by downloading a pre-compiled version, then symlinking the `bin` folder to where pyenv expects it.
+# In the future, we should consider dropping pyenv completely, as it's mostly used here for legacy reasons.
+
 FROM python-core as python-3.8
-RUN pyenv install $PY_3_8 \
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_8/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_8/x64/ \
+  && curl -L https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
+  | jq 'map(select(.version == '\"$PY_3_8\"')) [].files | map(select(.platform == "linux" and .arch == "x64" and .platform_version == "22.04"))[] | .download_url' \
+  | xargs -i -- curl -L {} \
+  | tar xz \
+  && rm build_output.txt Python-$PY_3_8.tgz tools_structure.txt setup.sh \
+  && mkdir /usr/local/.pyenv/versions/ \
+  && ln -s /opt/hostedtoolcache/Python/$PY_3_8/x64 /usr/local/.pyenv/versions/$PY_3_8 \
   && bash /opt/python/helpers/build $PY_3_8 \
-  && cd /usr/local/.pyenv \
-  && tar czf 3.8.tar.gz versions/$PY_3_8
+  && cd /opt/hostedtoolcache/Python \
+  && tar czf $PY_3_8.tar.gz $PY_3_8
 
 FROM python-core as python-3.9
-RUN pyenv install $PY_3_9 \
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_9/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_9/x64/ \
+  && curl -L https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
+  | jq 'map(select(.version == '\"$PY_3_9\"')) [].files | map(select(.platform == "linux" and .arch == "x64" and .platform_version == "22.04"))[] | .download_url' \
+  | xargs -i -- curl -L {} \
+  | tar xz \
+  && rm build_output.txt Python-$PY_3_9.tgz tools_structure.txt setup.sh \
+  && mkdir /usr/local/.pyenv/versions/ \
+  && ln -s /opt/hostedtoolcache/Python/$PY_3_9/x64 /usr/local/.pyenv/versions/$PY_3_9 \
   && bash /opt/python/helpers/build $PY_3_9 \
-  && cd /usr/local/.pyenv \
-  && tar czf 3.9.tar.gz versions/$PY_3_9
+  && cd /opt/hostedtoolcache/Python \
+  && tar czf $PY_3_9.tar.gz $PY_3_9
 
 FROM python-core as python-3.10
-RUN pyenv install $PY_3_10 \
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_10/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_10/x64/ \
+  && curl -L https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
+  | jq 'map(select(.version == '\"$PY_3_10\"')) [].files | map(select(.platform == "linux" and .arch == "x64" and .platform_version == "22.04"))[] | .download_url ' \
+  | xargs -i -- curl -L {} \
+  | tar xz \
+  && rm build_output.txt Python-$PY_3_10.tgz tools_structure.txt setup.sh \
+  && mkdir /usr/local/.pyenv/versions/ \
+  && ln -s /opt/hostedtoolcache/Python/$PY_3_10/x64 /usr/local/.pyenv/versions/$PY_3_10 \
   && bash /opt/python/helpers/build $PY_3_10 \
-  && cd /usr/local/.pyenv \
-  && tar czf 3.10.tar.gz versions/$PY_3_10
+  && cd /opt/hostedtoolcache/Python \
+  && tar czf $PY_3_10.tar.gz $PY_3_10
```

Comment:
I hate the lack of DRY across all these, but AFAIK there's no clean way to loop something like this in a multi-stage docker build.

I didn't duplicate the comments across each one... decided that doing it solely on the default `3.11` install sufficed.